microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java

```diff
@@ -51,84 +45,66 @@
+        addHealthCheck(ReadinessLiteral.INSTANCE, bm, reporter::addReadinessCheck);
     }
 
-    private void addHealthCheck(BeanManager bm,
-                                List<AnnotatedType<? extends HealthCheck>> delegates,
-                                Collection<Unmanaged.UnmanagedInstance<HealthCheck>> instances,
+    private void addHealthCheck(AnnotationLiteral qualifier, BeanManager bm,
                                 Consumer<HealthCheck> healthFunction) {
-        for (AnnotatedType delegate : delegates) {
-            try {
-                Unmanaged<HealthCheck> unmanagedHealthCheck = new Unmanaged<HealthCheck>(bm, delegate.getJavaClass());
-                Unmanaged.UnmanagedInstance<HealthCheck> healthCheckInstance = unmanagedHealthCheck.newInstance();
-                HealthCheck healthCheck = healthCheckInstance.produce().inject().postConstruct().get();
-                instances.add(healthCheckInstance);
-                healthFunction.accept(healthCheck);
-            } catch (Exception e) {
-                throw new RuntimeException("Failed to register health bean", e);
-            }
+        Instance<HealthCheck> healthChecks = bm.createInstance().select(HealthCheck.class, qualifier);
+        for (HealthCheck healthCheck : healthChecks) {
+            healthFunction.accept(healthCheck);
         }
     }
 
     /**
      * Called when the deployment is undeployed.
      * <p>
-     * Remove all the instances of {@link HealthCheck} from the {@link SmallRyeHealthReporter}.
-     * Handle manually their CDI destroy lifecycle.
+     * Remove all the instances of {@link HealthCheck} from the {@link HealthReporter}.
      */
-    public void beforeShutdown(@Observes final BeforeShutdown bs) {
-        removeHealthCheck(healthInstances, reporter::removeHealthCheck);
-        removeHealthCheck(readinessInstances, reporter::removeReadinessCheck);
-        removeHealthCheck(livenessInstances, reporter::removeLivenessCheck);
+    public void beforeShutdown(@Observes final BeforeShutdown bs, BeanManager bm) {
+        removeHealthCheck(HealthLiteral.INSTANCE, bm, reporter::removeHealthCheck);
+        removeHealthCheck(LivenessLiteral.INSTANCE, bm, reporter::removeLivenessCheck);
+        removeHealthCheck(ReadinessLiteral.INSTANCE, bm, reporter::removeReadinessCheck);
     }
 
-    private void removeHealthCheck(Collection<Unmanaged.UnmanagedInstance<HealthCheck>> instances,
+    private void removeHealthCheck(AnnotationLiteral qualifier, BeanManager bm,
                                    Consumer<HealthCheck> healthFunction) {
-        instances.forEach(healthCheck -> {
-            healthFunction.accept(healthCheck.get());
-            healthCheck.preDestroy().dispose();
-        });
-        instances.clear();
+        Instance<HealthCheck> healthChecks = bm.createInstance().select(HealthCheck.class, qualifier);
+        for (HealthCheck healthCheck : healthChecks) {
+            healthFunction.accept(healthCheck);
```

Comment:
Note that every invocation of `get()` on an `Instance` results in *new* injection which in case of `@Dependent` bean means a whole new bean. So you would be attempting to remove health check on different beans then where you added it.

Best way would be to store the set of beans the first time you got them in `addHealthCheck()` and then in `removeHealthCheck()` you would just iterate over them and finally do `Instance.destroy()` on all of them (since it happens before shutdown, it should be safe)?